specification.md

```diff
@@ -0,0 +1,92 @@
+
+#### Involved Contracts
+- `PreImageOracle`: It is my understanding that this is interacted with the solver and not our agents (need to check with Ben)
+- `DisputeGameFactory`: This contract is used to create dispute game instances which enable the dispute game to be played. The game is played when players submit claims. A Claim is only submitted in opposition to the current state. 
+- `MIPs`: MIPs is the mips32 instruction set architecture in solidity. This is what produces a cannon trace which is the state of L2 that players make claims about.
+
+#### Agents
+
+Agents are defined by their behaviors. We define a behavior trait as follows
+```rust
+/// The [`Behavior`] trait is the lowest-level functionality that will be used
+/// by a [`StateMachine`]. This constitutes what each state transition will do.
+#[async_trait::async_trait]
+pub trait Behavior<E>: Send + Sync + 'static {
+    /// Used to bring the agent back up to date with the latest state of the
+    /// world. This could be used if the world was stopped and later restarted.
+    async fn sync(&mut self, _messager: Messager, _client: Arc<RevmMiddleware>) {}
+
+    /// Used to start the agent.
+    /// This is where the agent can engage in its specific start-up activities
+    /// that it can do, given the current state of the world.
+    async fn startup(&mut self) {}
+
+    /// Used to process events.
+    /// This is where the agent can engage in its specific processing
+    /// of events that can lead to actions being taken.
+    async fn process(&mut self, event: E) -> Option<MachineHalt>;
+}
+```
+
+The dispute game simulation will have the following agents.
+
+- **Game Master:** Responsible for reading the state of a dispute game instance (created by the dispute game factory contract) and sending the new claims to Durin, the dispute game solver. Durin will return a correct move. The Game master will then send the correct move to all other agents (game players) in the simulation. The game master is the only agent who is not a player in the game. The Game master is the most complicated agent in the simulation. It will be responsible for the following:
+    - Reading the state of the dispute game instance
+    - Sending the state to Durin
+    - Receiving the correct move from Durin
+    - Sending the correct move to all other agents
+
+- **Honest Agent:** Responsible for making correct claims on the dispute game instance. This is the most straightforward agent in the simulation. It will be responsible for the following:
+    - Receiving the correct move from the game master
+    - Making a correct claim on a dispute game instance
+
+- **Malicious Agent:** Responsible for making incorrect claims on the dispute game instance. This agent is also quite simple but can have many different variations. It will be responsible for the following:
```

Comment:
I cannot emphasize enough how this is nontrivial. It's not just about posting something wrong, it's about trying to fuck with other agents to cause problems in their behavior. A simulation that shows a trivial malicious agent that just posts random state will not be hard, but this is Step 0.